csrc/executor_utils.cpp

```diff
@@ -407,11 +407,67 @@ void validateAlignedVectorizeExtents(
   //     info.word_size);
 }
 
+namespace {
+
+// Return offsets of the first points accessed. Currently only
+// non-zero when tensor is sliced.
+std::vector<size_t> getTensorOffsets(
```

Comment:
Two thing we are currently missing:
1. once we see a slice whose start is not 0, then this dim should no longer be considered as contiguous, therefore also no longer a inner dim.
2. once we see a slice whose stop is not extent, then the left dim should no longer be considered as contiguous.

I think to resolve the above, we have no choice but to  combine the slice check and stride check together. Something like:
```C++
for (all uses) {
  for (dim : dims inner to outer)
    auto contiguity = (dim is slice from nonzero or dim + 1 is slice to non-extent) ? false : tv->contiguity[dim];
    auto stride = is slice ? strides[dim] * start : strides[dim];
    switch (contiguity) {
    case true:
      check stride == cur_contig_stride 
      break;
    case null:
      if (is expanded broadcast) {
        check stride == 0
      } else {
        check is broadcast && size == 1;
      }
      break;
    case false:
      check stride % vec_size
    }
  }
}
```

Reply:
> once we see a slice whose start is not 0, then this dim should no longer be considered as contiguous, therefore also no longer a inner dim.

That doesn't seem to be true. We can certainly vectorize like:

```
t0 = [1025]
t1 = t0[0:1024]
```

> once we see a slice whose stop is not extent, then the left dim should no longer be considered as contiguous.

That is true, but non-zero start also makes the left dim non-contiguous, right?

Reply:
> That doesn't seem to be true. We can certainly vectorize like:

Isn't this a slice with zero start, which according to that statement, we do not consider it as discontiguous?

Reply:
> That is true, but non-zero start also makes the left dim non-contiguous, right?

I think non-zero start make this dim non-contiguous?

Reply:
> > That doesn't seem to be true. We can certainly vectorize like:
> 
> Isn't this a slice with zero start, which according to that statement, we do not consider it as discontiguous?

Sorry, I meant to have an example like:

```
t0 = [1025]
t1 = t0[1:]
```

As long as `&t0[1]` is aligned, we should be able vectorize this. 

Here's an example test: https://github.com/NVIDIA/Fuser/pull/733/files#diff-e827e7af168cc47583ab757606cafe080c3249256073de268c951040285cb67bR2485

Reply:
> Sorry, I meant to have an example like:

Oh, I see. For that case, I agree. Then I also agree with

> That is true, but non-zero start also makes the left dim non-contiguous

Reply:
I think I start to agree with you more.

For example, in:
```
t0: shape [1025, 4] stride [7, 1]
t1 = t0 // not vectorizable
t2 = t0[4:1024, :] // still not vectorizable
```
For this case although `stride[0] * 4` is a multiple of 4, we still can not vectorize to 4, because we are still accessing the item at `t0[5, :]` which is not aligned. For this case, the offset check and the stride check is basically independent.

Reply:
I am thinking about another example:
```C++
t0: shape [4, 1024, 3] // contiguous
t1 = t0 // vectorizable to 4
t2 = t0[:, 4:6, :] // vectorizable to 2
```
In this example, if my scheduler wrongly scheduled a vectorize 4 for t2, will I be able to catch this via validation? It seems not, because in the offset check, the start offset for t2 is `4*3`, which passes the check. The stride check also passes, because the whole `t0` is contiguous.

Reply:
Another similar example:
```C++
t0: shape [4, 1025, 3] // contiguous
t1 = t0 // vectorizable to 4
t2 = t0[:, 4:6, :] // not vectorizable
```
This is very similar to the above example, except the size of the second dim is changed from 1024 to 1025. For this case, `t2` is no longer vectorizable any more, because the stride of dim 0 is 3 * 1025, which can not guarantee that the 2x3 islands are always aligned.

So, indeed, having a slice does not change the contiguity of this dim, but it does make the left dim discontigous. So we should start checking the stride of the left dim, if we are not already doing so.

Reply:
> I think I start to agree with you more.
> 
> For example, in:
> 
> ```
> t0: shape [1025, 4] stride [7, 1]
> t1 = t0 // not vectorizable
> t2 = t0[4:1024, :] // still not vectorizable
> ```
> 
> For this case although `stride[0] * 4` is a multiple of 4, we still can not vectorize to 4, because we are still accessing the item at `t0[5, :]` which is not aligned. For this case, the offset check and the stride check is basically independent.

You're right. I think I need to update this part: https://github.com/NVIDIA/Fuser/pull/733/files#diff-7a4f65bea86e65f941636f4355972029ce48141c38d5ae2db39c10ad7a291207R495-R499. I wasn't paying attention to that as I didn't see any validation error in the original PR.

Reply:
> I am thinking about another example:
> 
> ```c++
> t0: shape [4, 1024, 3] // contiguous
> t1 = t0 // vectorizable to 4
> t2 = t0[:, 4:6, :] // vectorizable to 2
> ```
> 
> In this example, if my scheduler wrongly scheduled a vectorize 4 for t2, will I be able to catch this via validation? It seems not, because in the offset check, the start offset for t2 is `4*3`, which passes the check. The stride check also passes, because the whole `t0` is contiguous.

Yeah, this example seems to be hitting the case where the analysis here is not sufficient as it's not as comprehensive as the one in the scheduler. I'd say we shouldn't worry too much about it for now.

Reply:
> Another similar example:
> 
> ```c++
> t0: shape [4, 1025, 3] // contiguous
> t1 = t0 // vectorizable to 4
> t2 = t0[:, 4:6, :] // not vectorizable
> ```
> 
> This is very similar to the above example, except the size of the second dim is changed from 1024 to 1025. For this case, `t2` is no longer vectorizable any more, because the stride of dim 0 is 3 * 1025, which can not guarantee that the 2x3 islands are always aligned.
> 
> So, indeed, having a slice does not change the contiguity of this dim, but it does make the left dim discontigous. So we should start checking the stride of the left dim, if we are not already doing so.

Added a test case: `Slice3DVectorizeManual1`